### PR TITLE
core: mmu: protect global asid field with a lock

### PR DESCRIPTION
**Pull request by jenswi-linaro**
Protects the global ASID bitfield (g_asid) with a spinlock.

Fixes: 99f969dd6c99 ("core: fine grained tee_ta_mutex locking")
Signed-off-by: Jens Wiklander <jens.wiklander@linaro.org>
Tested-by: Jens Wiklander <jens.wiklander@linaro.org> (Hikey)

### CODE REVIEW

File: core/arch/arm/mm/tee_mmu.c [hunk 1]
```diff
@@ -227,27 +230,42 @@ static TEE_Result tee_mmu_umap_set_vas(struct tee_mmu_info *mmu)
 
 static unsigned int asid_alloc(void)
 {
+	uint32_t exceptions = cpu_spin_lock_xsave(&g_asid_spinlock);
+	unsigned int r = 0;
```
Comment (etienne-lms) [inline core/arch/arm/mm/tee_mmu.c]: minor: no need to init

File: core/arch/arm/mm/tee_mmu.c [hunk 2]
```diff
@@ -227,27 +230,42 @@ static TEE_Result tee_mmu_umap_set_vas(struct tee_mmu_info *mmu)
...
 	assert(!(asid & 1));
 
 	if (asid) {
 		int i = (asid - 1) / 2;
 
+		if (!(i < MMU_NUM_ASIDS && bit_test(g_asid, i))) {
+			EMSG("i %u bit %d", i, bit_test(g_asid, i));
+			panic();
+		}
 		assert(i < MMU_NUM_ASIDS && bit_test(g_asid, i));
```
Comment (etienne-lms) [inline core/arch/arm/mm/tee_mmu.c]: panic or assert?

File: core/arch/arm/mm/tee_mmu.c [hunk 3]
```diff
@@ -227,27 +230,42 @@ static TEE_Result tee_mmu_umap_set_vas(struct tee_mmu_info *mmu)
...
 {
+	uint32_t exceptions = cpu_spin_lock_xsave(&g_asid_spinlock);
+
 	/* Only even ASIDs are supposed to be allocated */
 	assert(!(asid & 1));
 
 	if (asid) {
 		int i = (asid - 1) / 2;
 
+		if (!(i < MMU_NUM_ASIDS && bit_test(g_asid, i))) {
```
Comment (etienne-lms) [inline core/arch/arm/mm/tee_mmu.c]: minor: more readable as `(i >= MMU_NUM_ASIDS || !bit_test(g_asid, i))`

Comment (jenswi-linaro) [reply core/arch/arm/mm/tee_mmu.c]: Sorry, this should be removed, it was just debug code.

**Comment by lorc** [comment]
`Reviewed-by: Volodymyr Babchuk <vlad.babchuk@gmail.com>`

File: core/arch/arm/mm/tee_mmu.c [hunk 4]
```diff
@@ -75,6 +76,8 @@
 #define MMU_NUM_ASIDS		64
 
 static bitstr_t bit_decl(g_asid, MMU_NUM_ASIDS);
+static unsigned int g_asid_spinlock = SPINLOCK_UNLOCK;
+
```
Comment (etienne-lms) [inline core/arch/arm/mm/tee_mmu.c]: minor: remove empty line.

File: core/arch/arm/mm/tee_mmu.c [hunk 5]
```diff
@@ -227,18 +230,27 @@ static TEE_Result tee_mmu_umap_set_vas(struct tee_mmu_info *mmu)
...
 
 	bit_ffc(g_asid, MMU_NUM_ASIDS, &i);
-	if (i == -1)
-		return 0;
-	bit_set(g_asid, i);
+	if (i == -1) {
+		r = 0;
+	} else {
+		bit_set(g_asid, i);
+
```
Comment (etienne-lms) [inline core/arch/arm/mm/tee_mmu.c]: minor: maybe remove this empty line too

**Comment by jenswi-linaro** [comment]
Rebased, squashed and tags applied

**Pull request merged**